tests/test_ty_check.rs

```diff
@@ -0,0 +1,1525 @@
+                y: self.y + other.y
+            }
+        }
+    }
+    fn main(x: &shrd gpu.global Point, y: &shrd gpu.global Point, res: &uniq gpu.global Point) -[gpu.thread]-> () {
+        let add = |x, y, res| -[gpu.thread]-> _ {
+            *res = *x + *y
+        };
+        let y = add(x, y, res)
+    }
+    "#;
+    assert_err_compile!(src);
+}
+
+#[test]
+#[ignore] // TODO Reassignment to dead types dont work
+fn test_struct_dead_types() {
+    let src = r#"
+    struct Test<a: prv, m: mem> {
+        test: &a uniq m i32
+    }
+    fn foo<m: mem>(a: &uniq m i32) -[cpu.thread]-> () {
+        ()
+    }
+    fn test() -[cpu.thread]-> () {
+        let mut x = 42;
+        let mut x2 = x;
+        let mut t = Test { test: &uniq x };
+        foo(t.test);
+        t.test = &uniq x2 //t.test is dead
+    }
+    "#;
+    assert_compile!(src);
+}
+
+#[test]
+fn test_lambda2() {
+    let src = r#"
+    struct Array<T, N: nat> {
+        data: [T; N]
+    }
+    impl<T, N: nat> Copy for Array<T, N> {}
+    impl<T, N: nat> Array<T, N> {
+        fn new_pair(t1: T, t2: T) -[gpu.thread]-> Array<T, 2> {
+            Array {
+                data: [t1, t2]
+            }
+        }
+    }
+    impl<T, N: nat, m: mem> Array<T, N> where T: Number {
+        fn reduce(&shrd m self, zero: T) -[gpu.thread]-> T {
+            let mut result = zero;
+            for_nat i in range(0, N) {
+                result = result + (*self).data[i]
+            };
+            result
+        }
+    }
+    fn main(array_global: &shrd gpu.global Array<i32, 2>) -[gpu.thread]-> () {
+        let red = |vec, zero| -[gpu.thread]-> _ {
+            vec.reduce(zero)
+        };
+        let array2 = Array<_, 2>::new_pair(16, 27);
+        let res = red(array_global, 0)
+    }
+    "#;
+    assert_compile!(src);
+
+    let src = r#"
+    struct Array<T, N: nat> {
+        data: [T; N]
+    }
+    impl<T, N: nat> Array<T, N> {
+        fn new_pair(t1: T, t2: T) -[gpu.thread]-> Array<T, 2> {
+            Array {
+                data: [t1, t2]
+            }
+        }
+    }
+    impl<T, N: nat, m: mem> Array<T, N> where T: Number {
+        fn reduce(&shrd m self, zero: T) -[gpu.thread]-> T {
+            let mut result = zero;
+            for_nat i in range(0, N) {
+                result = result + (*self).data[i]
+            };
+            result
+        }
+    }
+    fn main(array_global: &shrd gpu.global Array<i32, 2>) -[gpu.thread]-> () {
+        let red = |vec, zero| -[gpu.thread]-> _ {
+            vec.reduce(zero)
+        };
+        let array2 = Array<_, 2>::new_pair(16, 27);
+        let res = red(array_global, 0)
+    }
+    "#;
+    // Should this compile?
+    // It throws an error while typechecking `vec.reduce(zero)`
+    // First we typecheck `vec`. This adds the constraint that type of vec must implement Copy.
+    // So the is no suitable function with name `reduce` for a type which implements Copy
+    assert_err_compile!(src);
```

Comment:
Ein interessanter Testfall.
Falls `Array<T, N: nat>` nicht Copy ist, ist die Typinferenz nicht erfolgreich, weil in `vec.reduce(zero)` das Typechecking für `vec` ein Copy-constraint auf dem Typen von `vec` produziert